File: cia-frontend/packages/api-client/src/modules/policy.ts
// ── Policy — schemas + derived types ─────────────────────────────────────
//
// Mirrors cia-policy/dto/* records. Serves at /api/v1/policies/...
//
// Schemas are the source of truth — types are derived via z.infer.
// Use validated* helpers from '@cia/api-client' to fetch + validate:
//
//   import { validatedGet, PolicyDtoSchema } from '@cia/api-client';
//   const p = await validatedGet(`/api/v1/policies/${id}`, PolicyDtoSchema);

import { z } from 'zod';

// ── Enums ─────────────────────────────────────────────────────────────────

export const PolicyStatusSchema = z.enum([
  'DRAFT',
  'PENDING_APPROVAL',
  'ACTIVE',
  'REJECTED',
  'CANCELLED',
  'LAPSED',
  'EXPIRED',
  'REINSTATED',
]);
export type PolicyStatus = z.infer<typeof PolicyStatusSchema>;

export const BusinessTypeSchema = z.enum([
  'DIRECT',
  'DIRECT_WITH_COINSURANCE',
  'INWARD_COINSURANCE',
  'INWARD_FACULTATIVE',
]);
export type BusinessType = z.infer<typeof BusinessTypeSchema>;

export const SurveyStatusSchema = z.enum([
  'ASSIGNED',
  'REPORT_SUBMITTED',
  'APPROVED',
  'OVERRIDDEN',
]);
export type SurveyStatus = z.infer<typeof SurveyStatusSchema>;

// ── Risk ──────────────────────────────────────────────────────────────────

export const PolicyRiskDtoSchema = z.object({
  id:               z.string(),
  description:      z.string(),
  sumInsured:       z.number(),
  premium:          z.number(),
  sectionId:        z.string().nullable().optional(),
  sectionName:      z.string().nullable().optional(),
  riskDetails:      z.record(z.string(), z.unknown()).nullable().optional(),
  vehicleRegNumber: z.string().nullable().optional(),
  orderNo:          z.number(),
});
export type PolicyRiskDto = z.infer<typeof PolicyRiskDtoSchema>;

// ── Coinsurance participant ──────────────────────────────────────────────

export const PolicyCoinsuranceParticipantDtoSchema = z.object({
  id:                   z.string(),
  insuranceCompanyId:   z.string(),
  insuranceCompanyName: z.string(),
  sharePercentage:      z.number(),
});
export type PolicyCoinsuranceParticipantDto = z.infer<typeof PolicyCoinsuranceParticipantDtoSchema>;

// ── Survey (added in B4.3) ────────────────────────────────────────────────

export const PolicySurveyDtoSchema = z.object({
  id:                z.string(),
  policyId:          z.string(),
  status:            SurveyStatusSchema,

  surveyorType:      z.string().nullable().optional(),
  surveyorId:        z.string().nullable().optional(),
  surveyorName:      z.string().nullable().optional(),
  assignedBy:        z.string().nullable().optional(),
  assignedAt:        z.string().nullable().optional(),

  reportPath:        z.string().nullable().optional(),
  reportNotes:       z.string().nullable().optional(),
  reportUploadedBy:  z.string().nullable().optional(),
  reportUploadedAt:  z.string().nullable().optional(),

  approvedBy:        z.string().nullable().optional(),
  approvedAt:        z.string().nullable().optional(),
  approvalNotes:     z.string().nullable().optional(),

  overrideReason:    z.string().nullable().optional(),
  overriddenBy:      z.string().nullable().optional(),
  overriddenAt:      z.string().nullable().optional(),

  createdAt:         z.string(),
});
export type PolicySurveyDto = z.infer<typeof PolicySurveyDtoSchema>;

// Frozen clause snapshot on the policy (mirrors com.nubeero.cia.common.clause.ClauseSnapshot).
export const ClauseSnapshotDtoSchema = z.object({
  id:    z.string(),
  title: z.string(),
  text:  z.string(),
  type:  z.string(),
});
export type ClauseSnapshotDto = z.infer<typeof ClauseSnapshotDtoSchema>;

// ── Policy ────────────────────────────────────────────────────────────────

export const PolicyDtoSchema = z.object({
  id:                       z.string(),
  policyNumber:             z.string().nullable().optional(),
  status:                   PolicyStatusSchema,

  quoteId:                  z.string().nullable().optional(),
  quoteNumber:              z.string().nullable().optional(),

  customerId:               z.string(),
  customerName:             z.string(),

  productId:                z.string(),
  productName:              z.string(),
  productCode:              z.string(),
  productRate:              z.number(),

  classOfBusinessId:        z.string(),
  classOfBusinessName:      z.string(),
  classOfBusinessCode:      z.string(),

  brokerId:                 z.string().nullable().optional(),
  brokerName:               z.string().nullable().optional(),

  // Per-policy agent attribution (Slice 84d / V53). Mutually exclusive with
  // brokerId at the DB level (ck_policies_broker_xor_agent).
  agentId:                  z.string().nullable().optional(),
  agentName:                z.string().nullable().optional(),

  // Relationship Manager attribution (B2 Task 5.1). RM is an accrual-only
  // commission source (no payment/CreditNote) — surfaced on the Financial tab.
  relationshipManagerId:    z.string().nullable().optional(),
  relationshipManagerName:  z.string().nullable().optional(),

  businessType:             BusinessTypeSchema,
  niidRequired:             z.boolean(),

  policyStartDate:          z.string(),
  policyEndDate:            z.string(),

  totalSumInsured:          z.number(),
  totalPremium:             z.number(),
  discount:                 z.number(),
  netPremium:               z.number(),

  // Commission snapshot (Slice 84b V51 — surfaced by 84e). All three are null
  // when no commission is configured at issuance; paired-CHECK semantics on
  // the DB keep source + rate in lockstep. commissionAmount is computed by the
  // backend (netPremium × rate / 100, HALF_UP 2dp) so the frontend doesn't
  // multiply.
  commissionSourceType:     z.enum(['AGENT', 'BROKER', 'RELATIONSHIP_MANAGER']).nullable().optional(),
  commissionRate:           z.number().nullable().optional(),
  commissionAmount:         z.number().nullable().optional(),

  notes:                    z.string().nullable().optional(),
  workflowId:               z.string().nullable().optional(),

  approvedBy:               z.string().nullable().optional(),
  approvedAt:               z.string().nullable().optional(),
  rejectedBy:               z.string().nullable().optional(),
  rejectedAt:               z.string().nullable().optional(),
  rejectionReason:          z.string().nullable().optional(),

  cancelledBy:              z.string().nullable().optional(),
  cancelledAt:              z.string().nullable().optional(),
  cancellationReason:       z.string().nullable().optional(),

  naicomUid:                z.string().nullable().optional(),
  naicomUploadedAt:         z.string().nullable().optional(),
  naicomCertificatePath:    z.string().nullable().optional(),

  niidRef:                  z.string().nullable().optional(),
  niidUploadedAt:           z.string().nullable().optional(),

  policyDocumentPath:       z.string().nullable().optional(),
  documentSentAt:           z.string().nullable().optional(),
  documentSentBy:           z.string().nullable().optional(),
  documentAcknowledgedAt:   z.string().nullable().optional(),
  documentAcknowledgedBy:   z.string().nullable().optional(),

  risks:                    z.array(PolicyRiskDtoSchema),
  coinsuranceParticipants:  z.array(PolicyCoinsuranceParticipantDtoSchema),
  selectedClauses:          z.array(ClauseSnapshotDtoSchema).optional(),
  survey:                   PolicySurveyDtoSchema.nullable().optional(),

  createdAt:                z.string(),
  updatedAt:                z.string().nullable().optional(),
});
export type PolicyDto = z.infer<typeof PolicyDtoSchema>;

// ── Policy summary (list endpoint returns this lighter shape) ─────────────

export const PolicySummaryDtoSchema = z.object({
  id:                  z.string(),
  policyNumber:        z.string().nullable().optional(),
  status:              PolicyStatusSchema,
  customerId:          z.string(),
  customerName:        z.string(),
  productName:         z.string(),
  classOfBusinessName: z.string(),
  brokerName:          z.string().nullable().optional(),
  /** Per-policy agent attribution (Slice 84d). Mutually exclusive with brokerName via V53 CHECK. */
  agentName:           z.string().nullable().optional(),
  businessType:        BusinessTypeSchema,
  policyStartDate:     z.string(),
  policyEndDate:       z.string(),
  totalSumInsured:     z.number(),
  netPremium:          z.number(),
  naicomUid:           z.string().nullable().optional(),
  createdAt:           z.string(),
});
export type PolicySummaryDto = z.infer<typeof PolicySummaryDtoSchema>;

File: cia-backend/cia-policy/src/main/java/com/nubeero/cia/policy/Policy.java
package com.nubeero.cia.policy;

import com.nubeero.cia.common.entity.BaseEntity;
import com.nubeero.cia.quotation.BusinessType;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "policies")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Policy extends BaseEntity {

    @Column(name = "policy_number", unique = true, length = 60)
    private String policyNumber;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    @Builder.Default
    private PolicyStatus status = PolicyStatus.DRAFT;

    // ── Quote linkage ────────────────────────────────────────────────────
    @Column(name = "quote_id")
    private UUID quoteId;

    @Column(name = "quote_number", length = 30)
    private String quoteNumber;

    // ── Customer snapshot ────────────────────────────────────────────────
    @Column(name = "customer_id", nullable = false)
    private UUID customerId;

    @Column(name = "customer_name", nullable = false, length = 200)
    private String customerName;

    // ── Product snapshot ─────────────────────────────────────────────────
    @Column(name = "product_id", nullable = false)
    private UUID productId;

    @Column(name = "product_name", nullable = false, length = 100)
    private String productName;

    @Column(name = "product_code", nullable = false, length = 20)
    private String productCode;

    @Column(name = "product_rate", nullable = false, precision = 10, scale = 4)
    private BigDecimal productRate;

    // ── Class of business snapshot ───────────────────────────────────────
    @Column(name = "class_of_business_id", nullable = false)
    private UUID classOfBusinessId;

    @Column(name = "class_of_business_name", nullable = false, length = 100)
    private String classOfBusinessName;

    @Column(name = "class_of_business_code", nullable = false, length = 20)
    private String classOfBusinessCode;

    // ── Broker ───────────────────────────────────────────────────────────
    @Column(name = "broker_id")
    private UUID brokerId;

    @Column(name = "broker_name", length = 100)
    private String brokerName;

    @Enumerated(EnumType.STRING)
    @Column(name = "business_type", nullable = false, length = 30)
    @Builder.Default
    private BusinessType businessType = BusinessType.DIRECT;

    @Column(name = "niid_required", nullable = false)
    @Builder.Default
    private boolean niidRequired = false;

    @Column(name = "policy_start_date", nullable = false)
    private LocalDate policyStartDate;

    @Column(name = "policy_end_date", nullable = false)
    private LocalDate policyEndDate;

    // ── Financials ───────────────────────────────────────────────────────
    @Column(name = "total_sum_insured", nullable = false, precision = 18, scale = 2)
    @Builder.Default
    private BigDecimal totalSumInsured = BigDecimal.ZERO;

    @Column(name = "total_premium", nullable = false, precision = 18, scale = 2)
    @Builder.Default
    private BigDecimal totalPremium = BigDecimal.ZERO;

    @Column(nullable = false, precision = 18, scale = 2)
    @Builder.Default
    private BigDecimal discount = BigDecimal.ZERO;

    @Column(name = "net_premium", nullable = false, precision = 18, scale = 2)
    @Builder.Default
    private BigDecimal netPremium = BigDecimal.ZERO;

    @Column(columnDefinition = "TEXT")
    private String notes;

    @Column(name = "workflow_id", length = 200)
    private String workflowId;

    // ── Approval ─────────────────────────────────────────────────────────
    @Column(name = "approved_by", length = 100)
    private String approvedBy;

    @Column(name = "approved_at")
    private Instant approvedAt;

    @Column(name = "rejected_by", length = 100)
    private String rejectedBy;

    @Column(name = "rejected_at")
    private Instant rejectedAt;

    @Column(name = "rejection_reason", columnDefinition = "TEXT")
    private String rejectionReason;

    // ── Cancellation ─────────────────────────────────────────────────────
    @Column(name = "cancelled_by", length = 100)
    private String cancelledBy;

    @Column(name = "cancelled_at")
    private Instant cancelledAt;

    @Column(name = "cancellation_reason", columnDefinition = "TEXT")
    private String cancellationReason;

    // ── NAICOM ───────────────────────────────────────────────────────────
    @Column(name = "naicom_uid", length = 100)
    private String naicomUid;

    @Column(name = "naicom_uploaded_at")
    private Instant naicomUploadedAt;

    @Column(name = "naicom_certificate_path", length = 500)
    private String naicomCertificatePath;

    // ── NIID ─────────────────────────────────────────────────────────────
    @Column(name = "niid_ref", length = 100)
    private String niidRef;

    @Column(name = "niid_uploaded_at")
    private Instant niidUploadedAt;

    // ── Document paths ───────────────────────────────────────────────────
    @Column(name = "policy_document_path", length = 500)
    private String policyDocumentPath;

    // ── Document delivery / acknowledgement ─────────────────────────────
    @Column(name = "document_sent_at")
    private Instant documentSentAt;

    @Column(name = "document_sent_by", length = 100)
    private String documentSentBy;

    @Column(name = "document_acknowledged_at")
    private Instant documentAcknowledgedAt;

    @Column(name = "document_acknowledged_by", length = 100)
    private String documentAcknowledgedBy;

    // ── Relationships ─────────────────────────────────────────────────────
    @OneToMany(mappedBy = "policy", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("orderNo ASC")
    @Builder.Default
    private List<PolicyRisk> risks = new ArrayList<>();

    @OneToMany(mappedBy = "policy", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<PolicyCoinsuranceParticipant> coinsuranceParticipants = new ArrayList<>();
}
